File: projects/zig/zig-on-the-web/src/window.zig
// ------ language="Zig" file="projects/zig/zig-on-the-web/src/window.zig" project://article.md#412
extern fn _alert(msg: [*]const u8, len: usize) void;

pub fn alert(msg: []u8) void {
  _alert(msg, msg.len);
}
// ------ end

File: projects/zig/zig-on-the-web/src/window.js
// ------ language="JavaScript" file="projects/zig/zig-on-the-web/src/window.js" project://article.md#447
const getString = (p, len) =>
  (new TextDecoder()).decode(app.exports.memory.buffer.slice(p, p + len));

var app = {
  imports: {
    window: {
      alert: (p, len) => alert(getString(p, len)),
    },
  },
  launch: r => {
    app.exports = r.instance.exports;
    if (!app.exports.launch()) throw "Launch Error";
  },
  exports: undefined,
};
// ------ end
